File: workshop/data/ZavaData.cs
// Shared ZavaShop data loader for every .NET LAB.
//
// Mirrors workshop/data/zava_data.py. Each loader returns plain
// System.Text.Json.Nodes.JsonNode trees (or convenience records for the
// most-used rows) so the JSON fixtures stay the single source of truth
// across the Python and .NET tracks.
//
// Reference it from a LAB csproj with:
//   <ItemGroup>
//     <Compile Include="..\data\ZavaData.cs" Link="ZavaData.cs" />
//   </ItemGroup>
//
// Then in any executable: `using ZavaShop.Workshop.Data;` and
// `ZavaData.FindStock("SKU-7421", "SEA-01")`.

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace ZavaShop.Workshop.Data;

public static class ZavaData
{
    private static readonly Lazy<string> _dataDir = new(() =>
    {
        // Walk up from the executable until we find the workshop/data folder.
        var dir = new DirectoryInfo(AppContext.BaseDirectory);
        while (dir is not null)
        {
            var candidate = Path.Combine(dir.FullName, "workshop", "data");
            if (Directory.Exists(candidate) && File.Exists(Path.Combine(candidate, "warehouses.json")))
            {
                return candidate;
            }

            // Or sibling: when the LAB binary runs from workshop/LABxx/bin/...
            var sibling = Path.Combine(dir.FullName, "..", "data");
            sibling = Path.GetFullPath(sibling);
            if (Directory.Exists(sibling) && File.Exists(Path.Combine(sibling, "warehouses.json")))
            {
                return sibling;
            }

            dir = dir.Parent;
        }

        throw new DirectoryNotFoundException(
            "Could not locate workshop/data — run the LAB from inside the repo.");
    });

    private static readonly Dictionary<string, JsonArray> _cache = new();

    public static string DataDir => _dataDir.Value;

    public static JsonArray Load(string fileName)
    {
        if (_cache.TryGetValue(fileName, out var cached))
        {
            return cached;
        }

        var path = Path.Combine(DataDir, fileName);
        var json = File.ReadAllText(path);
        var node = JsonNode.Parse(json)
            ?? throw new JsonException($"Failed to parse {fileName}");
        var array = node.AsArray();
        _cache[fileName] = array;
        return array;
    }

    public static IEnumerable<JsonNode?> LoadJsonLines(string fileName)
    {
        var path = Path.Combine(DataDir, fileName);
        foreach (var line in File.ReadAllLines(path))
        {
            if (string.IsNullOrWhiteSpace(line))
            {
                continue;
            }

            yield return JsonNode.Parse(line);
        }
    }

    // ---------- Loaders (full collection) ----------

    public static JsonArray LoadWarehouses() => Load("warehouses.json");
    public static JsonArray LoadSkus() => Load("skus.json");
    public static JsonArray LoadInventory() => Load("inventory.json");
    public static JsonArray LoadPurchaseOrders() => Load("purchase_orders.json");
    public static JsonArray LoadSuppliers() => Load("suppliers.json");
    public static JsonArray LoadContracts() => Load("contracts.json");
    public static JsonArray LoadCustomers() => Load("customers.json");
    public static JsonArray LoadOrders() => Load("orders.json");
    public static JsonArray LoadCarriers() => Load("carriers.json");
    public static JsonArray LoadExceptions() => Load("exceptions.json");
    public static IEnumerable<JsonNode?> LoadEvalQueries() => LoadJsonLines("eval_queries.jsonl");

    // ---------- Finders (return one row by id, or null) ----------

    public static JsonNode? FindStock(string sku, string warehouse) =>
        LoadInventory().FirstOrDefault(row =>
            row?["sku"]?.GetValue<string>() == sku &&
            row?["warehouse"]?.GetValue<string>() == warehouse);

    public static JsonNode? FindPo(string poNumber) =>
        LoadPurchaseOrders().FirstOrDefault(row =>
            row?["po_number"]?.GetValue<string>() == poNumber);

    public static JsonNode? FindSupplier(string supplierIdOrName)
    {
        var needle = supplierIdOrName.ToLowerInvariant();
        return LoadSuppliers().FirstOrDefault(row =>
            row?["supplier_id"]?.GetValue<string>()?.ToLowerInvariant() == needle ||
            row?["name"]?.GetValue<string>()?.ToLowerInvariant() == needle);
    }

    public static JsonNode? FindContract(string supplierId) =>
        LoadContracts().FirstOrDefault(row =>
            row?["supplier_id"]?.GetValue<string>() == supplierId);

    public static JsonNode? FindCustomer(string customerId) =>
        LoadCustomers().FirstOrDefault(row =>
            row?["customer_id"]?.GetValue<string>() == customerId);

    public static JsonNode? FindOrder(string orderId) =>
        LoadOrders().FirstOrDefault(row =>
            row?["order_id"]?.GetValue<string>() == orderId);
}

File: workshop/data/zava_data.py
"""
Shared ZavaShop data loader used by every LAB.

All LAB scripts import from this module so they share the same fixture set
under `workshop/data/`. Each loader returns plain Python lists / dicts —
no SDK types — so the data is reusable from function tools, evaluation
inputs, AG-UI server payloads, and workflow executors alike.
"""

from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path
from typing import Any

DATA_DIR = Path(__file__).parent


def _read_json(name: str) -> Any:
    return json.loads((DATA_DIR / name).read_text(encoding="utf-8"))


def _read_jsonl(name: str) -> list[dict[str, Any]]:
    path = DATA_DIR / name
    return [json.loads(line) for line in path.read_text(encoding="utf-8").splitlines() if line.strip()]


@lru_cache(maxsize=1)
def load_warehouses() -> list[dict[str, Any]]:
    return _read_json("warehouses.json")


@lru_cache(maxsize=1)
def load_skus() -> list[dict[str, Any]]:
    return _read_json("skus.json")


@lru_cache(maxsize=1)
def load_inventory() -> list[dict[str, Any]]:
    return _read_json("inventory.json")


@lru_cache(maxsize=1)
def load_purchase_orders() -> list[dict[str, Any]]:
    return _read_json("purchase_orders.json")


@lru_cache(maxsize=1)
def load_suppliers() -> list[dict[str, Any]]:
    return _read_json("suppliers.json")


@lru_cache(maxsize=1)
def load_contracts() -> list[dict[str, Any]]:
    return _read_json("contracts.json")


@lru_cache(maxsize=1)
def load_customers() -> list[dict[str, Any]]:
    return _read_json("customers.json")


@lru_cache(maxsize=1)
def load_orders() -> list[dict[str, Any]]:
    return _read_json("orders.json")


@lru_cache(maxsize=1)
def load_carriers() -> list[dict[str, Any]]:
    return _read_json("carriers.json")


@lru_cache(maxsize=1)
def load_exceptions() -> list[dict[str, Any]]:
    return _read_json("exceptions.json")


@lru_cache(maxsize=1)
def load_eval_queries() -> list[dict[str, Any]]:
    return _read_jsonl("eval_queries.jsonl")


# --------- small convenience lookups (used across labs) --------- #


def find_stock(sku: str, warehouse: str) -> dict[str, Any] | None:
    for row in load_inventory():
        if row["sku"] == sku and row["warehouse"] == warehouse:
            return row
    return None


def find_po(po_number: str) -> dict[str, Any] | None:
    for row in load_purchase_orders():
        if row["po_number"] == po_number:
            return row
    return None


def find_supplier(supplier_id_or_name: str) -> dict[str, Any] | None:
    needle = supplier_id_or_name.lower()
    for row in load_suppliers():
        if row["supplier_id"].lower() == needle or row["name"].lower() == needle:
            return row
    return None


def find_contract(supplier_id: str) -> dict[str, Any] | None:
    for row in load_contracts():
        if row["supplier_id"] == supplier_id:
            return row
    return None


def find_customer(customer_id: str) -> dict[str, Any] | None:
    for row in load_customers():
        if row["customer_id"] == customer_id:
            return row
    return None


def find_order(order_id: str) -> dict[str, Any] | None:
    for row in load_orders():
        if row["order_id"] == order_id:
            return row
    return None
